src/map.js

```diff
@@ -62,8 +65,8 @@ function map(object, f) {
         n = object.length,
         o;
 
-    if (f == null) while (++i < n) map.set(i, object[i]);
-    else while (++i < n) map.set(f(o = object[i], i, object), o);
+    if (f == null) while (++i < n) map.set(i, g(o = object[i], i, object));
+    else while (++i < n) map.set(f(o = object[i], i, object), g(o, i, object));
```

Comment:
Since *map* may be used widely in production in all kinds of situations, it may be worth optimizing to avoid the extra unnecessary function call to the identity function here.

This would require writing separate code for each of the 4 cases, and would eliminate the need for the `if(g == null) g = function(d) { return d }` step above:

```js
if (!f && !g)
```

```js
if (f && !g)
```

```js
if (!f && g)
```

```js
if (f && g)
```

Just a thought, not sure if it's worth it.